### DISCUSSION TITLE
Jumbled syntax tree with injections

### DISCUSSION THREAD

**Discussion by goerz** [nvim-treesitter/nvim-treesitter#7377] [Q&A]
I'm looking at [a Julia file](https://github.com/JuliaQuantumControl/QuantumPropagators.jl/blob/a0f7a46ab56dc70ccbdd1f42ef48641ecbd8af4e/src/propagate.jl) using the [`injections.scm`](https://github.com/nvim-treesitter/nvim-treesitter/blob/9e66d5c72dabfc6417c6015ed8991f3c4b504f7d/queries/julia/injections.scm) that ships with `nvim-treesitter`. This controls the injection of the `markdown` parser into Julia docstrings.

However, the syntax tree for the injected `markdown` appears to be jumbled:

![CleanShot 2024-11-19 at 21 10 12](https://github.com/user-attachments/assets/a4252981-929d-4e67-9cf3-27587f920908)

Note how I'm looking at a docstring at line 379 (380 in the 1-based vim UI), but the syntax tree contains an `inline` element originating from line 51, which is some random piece of text from another docstring. What's up with that? @clason

**Comment by clason** [comment]
AFAICT, this is just how the tree (of trees) is serialized for printing. If you can create a minimal(!) example that is obviously wrong, please open an issue at neovim/neovim (which is responsible for all this; we just install the parser and queries).

**Reply by goerz** [reply]
Huh, okay… that's kinda weird. If I can't trust that the apparent tree structure that I see in `:InspectTree` matches the actual structure in the parser, that makes it a bit hard to define queries.

I was wondering if maybe there was some aspect of the query syntax that I didn't get (I only learned the syntax this week), or some aspect of how injections work, where injection somehow pulls together nodes from siblings. This was in the context of trying to figure out whether these jumbled trees were somehow related to the problem in #7375.

But yeah, I should probably trim this down to a MWE and check in with the main neovim repo

**Reply by clason** [reply]
A minimal example would definitely help; GIFs are surprisingly bad at conveying (a lot of) context. And it's important to keep in mind that what you're seeing is not _one_ tree (which _does_ match the "structure in the(!) parser" but a linearization of several trees, each of which coming from a parser. There could be an issue with how these are interleaved, but it's hard to tell here.

**Reply by goerz** [reply]
> but a linearization of several trees

That's interesting… Can you point me to some docs that explain that in more detail, by any chance? I thought when I use an injection, it just takes the node that matches the injection query, and replaces its children with the tree generated by the injected parser. Why are there several trees involved?

At this point, I'm almost certain the problem I'm having in #7375 is, in fact, connected to this issue here: the inclusion I've defined for raw doc strings (`@doc raw"""..."""`) does actually show up as `markdown`/`markdown_inline` nodes in my `InspectTree`, but those nodes are in random places in the tree. It seems I'm just not seeing the effect (proper syntax highlighting) because of that, and I didn't initially realize there are new nodes in the tree.

**Reply by clason** [reply]
>  I thought when I use an injection, it just takes the node that matches the injection query, and replaces its children with the tree generated by the injected parser. Why are there several trees involved?

No, that's not true at all. Each tree is "owned" by a parser; you cannot mix-and-match. Instead, there's a higher-level object ("language tree") that tracks which regions of the buffer are parsed by which tree.

Documentation is lacking (WIP), but see `:h vim.treesitter.languagetree`.

> At this point, I'm almost certain the problem I'm having in https://github.com/nvim-treesitter/nvim-treesitter/discussions/7375 is, in fact, connected to this issue here:

That could very well be the case; markdown is tricky because of the nested parser. (Luckily, it's bundled with Neovim so unless the issue is specific to the Julia parser -- which may be the case; it's not the simplest -- you should be able to reproduce this from `nvim --clean`.)

In any case, this is a Neovim (or tree-sitter-julia) issue; nothing that can be solved here.

**Comment by clason** [comment]
Please try the latest `master`; the injection queries should be fixed now. (Not sure about the tree view, though.)
